cocotb/queue.py

```diff
@@ -0,0 +1,148 @@
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+
+import collections
+import heapq
+from asyncio import QueueEmpty, QueueFull
+
+import cocotb
+from cocotb.triggers import Event
+
+
+class Queue:
+    def __init__(self, maxsize=0):
+
+        self._maxsize = maxsize
+
+        self._finished = Event()
+        self._finished.set()
+
+        self._getters = collections.deque()
+        self._putters = collections.deque()
+
+        self._unfinished_tasks = 0
+
+        self._init(maxsize)
+
+    def _init(self, maxsize):
+        self._queue = collections.deque()
+
+    def _put(self, item):
+        self._queue.append(item)
+
+    def _get(self):
+        return self._queue.popleft()
+
+    def _wakeup_next(self, waiters):
+        while waiters:
+            event, task = waiters.popleft()
+            if not task._finished:
+                event.set()
+                break
+
+    def __repr__(self):
+        return '<{} at {:#x} {}>'.format(type(self).__name__, id(self), self._format())
+
+    def __str__(self):
+        return '<{} {}>'.format(type(self).__name__, self._format())
+
+    def __class_getitem__(cls, type):
+        return cls
+
+    def _format(self):
+        result = 'maxsize={}'.format(repr(self._maxsize))
+        if getattr(self, '_queue', None):
+            result += ' _queue={}'.format(repr(list(self._queue)))
+        if self._getters:
+            result += ' _getters[{}]'.format(len(self._getters))
+        if self._putters:
+            result += ' _putters[{}]'.format(len(self._putters))
+        if self._unfinished_tasks:
+            result += ' tasks={}'.format(self._unfinished_tasks)
+        return result
+
+    def qsize(self):
+        return len(self._queue)
+
+    @property
+    def maxsize(self):
+        return self._maxsize
+
+    def empty(self):
+        return not self._queue
+
+    def full(self):
+        if self._maxsize <= 0:
+            return False
+        else:
+            return self.qsize() >= self._maxsize
+
+    async def put(self, item):
+        while self.full():
+            event = Event()
```

Comment:
Having named Events would help with debug.
```suggestion
            event = Event('{} put'.format(type(self).__name__))
```